pkg/csi/service/vanilla/controller.go

```diff
@@ -1354,10 +1360,198 @@ func (c *controller) ValidateVolumeCapabilities(ctx context.Context, req *csi.Va
+		if err != nil {
+			return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to get nodes(node vms) in the vanilla cluster. Error: %v", err)
+		}
+
+		nextToken := ""
+		endIndex := maxEntries + startingToken
+		if endIndex > len(cnsQueryResult.Volumes) {
+			endIndex = len(cnsQueryResult.Volumes)
+		}
+		log.Debugf("Starting token: %d, End index: %d, Length of Query volume result: %d, Max entries: %d ",
+			startingToken, endIndex, len(cnsQueryResult.Volumes), maxEntries)
+		entries, nextToken, volumeType, err = c.processQueryResultsListVolumes(ctx, startingToken, endIndex,
+			cnsQueryResult, allNodeVMs)
+		if err != nil {
+			return nil, csifault.CSIInternalFault, fmt.Errorf("error while processing query results for list "+
+				" volumes, err: %v", err)
+		}
+		resp := &csi.ListVolumesResponse{
+			Entries:   entries,
+			NextToken: nextToken,
+		}
+
+		log.Infof("ListVolumes served %d results, token for next set: %s", len(entries), nextToken)
+		return resp, "", nil
+	}
+	listVolResponse, faultType, err := listVolumesInternal()
+	log.Debugf("List volume response: %+v", listVolResponse)
+	if err != nil {
+		prometheus.CsiControlOpsHistVec.WithLabelValues(volumeType, prometheus.PrometheusListVolumeOpType,
+			prometheus.PrometheusFailStatus, faultType).Observe(time.Since(start).Seconds())
+	} else {
+		prometheus.CsiControlOpsHistVec.WithLabelValues(volumeType, prometheus.PrometheusListVolumeOpType,
+			prometheus.PrometheusPassStatus, faultType).Observe(time.Since(start).Seconds())
+	}
+	return listVolResponse, err
+}
+
+func (c *controller) processQueryResultsListVolumes(ctx context.Context, startingToken int, endIndex int,
+	queryResult *cnstypes.CnsQueryResult, allNodeVMs []*cnsvsphere.VirtualMachine) ([]*csi.ListVolumesResponse_Entry,
+	string, string, error) {
+
+	volumeType := ""
+	nextToken := ""
+	log := logger.GetLogger(ctx)
+	var entries []*csi.ListVolumesResponse_Entry
+
+	volumeIDToNodeUUIDMap, err := c.getBlockVolumeToHostMap(ctx, allNodeVMs)
+	if err != nil {
+		return entries, nextToken, volumeType, err
+	}
+
+	for i := startingToken; i < endIndex; i++ {
+		if queryResult.Volumes[i].VolumeType == common.FileVolumeType {
+			volumeType = prometheus.PrometheusFileVolumeType
+			fileVolID := queryResult.Volumes[i].VolumeId.Id
+
+			// Populate csi.Volume info for the given volume
+			fileVolumeInfo := &csi.Volume{
+				VolumeId: fileVolID,
+			}
+
+			// Getting published nodes
+			publishedNodeIds := commonco.ContainerOrchestratorUtility.GetNodesForVolumes(ctx, []string{fileVolID})
+			nodeVMObj, err := c.nodeMgr.GetNodeByName(ctx, publishedNodeIds[fileVolID][0])
+			if err != nil {
+				log.Errorf("Failed to get node vm object from the node name, err:%v", err)
+				return entries, nextToken, volumeType, err
+			}
+			nodeVMUUID := nodeVMObj.UUID
+			log.Debugf("UUID of the node on which file volume ID: %s, is mounted: %s", fileVolID, nodeVMUUID)
```

Comment:
This will dump too many logs at debug for every minute. 
Think about wrapping all info at once via a single debug log.